Add FPCA warning to agreement signing page. Fixes: #1530

Jkhall81 opened the pull request:
Add FPCA warning to agreement signing page.

Fixes: #1530

Jkhall81 commented:
Hi! I'm new to contributing to noggin. 

This PR adds a warning to the agreement signing page when users haven't signed the FPCA, addressing issue #1530. The warning explains that not signing will prevent login to other Fedora services.

I placed the warning in `registration-agreements.html` since that's the page shown in the issue screenshot. Please let me know if:
1. The warning text needs adjustment
2. It should also go somewhere else (like during authentication)
3. There are any other changes needed

Thanks

abompard commented on noggin/templates/registration-agreements.html:
```diff
@@ -13,6 +12,25 @@
+                  {{ _("Not signing the FPCA will prevent you from logging in to many Fedora services such as Pagure, Fedora Discussion, and other contributor platforms.") }}
+                </div>
+              {% endif %}
+              <!-- END FPCA WARNING -->
```
You can move this block in the `_agreements_form.html` file, where there is already a loop over the agreements.

This way you don't need the block above where you set the `fpca` variable.

Jkhall81 replied:
Thanks for the feedback! I moved the FPCA warning block into _agreements_form.html as suggested and removed the logic from registration-agreements.html.

I also tried setting up a full Noggin development environment so I could walk through the UI changes directly, but I wasn’t able to get the user agreement flow working in my test setup. I’ll keep improving my local setup, but in the meantime I relied on reading the templates and following your guidance.

Please let me know if anything else should be adjusted!

abompard reviewed (changes requested):
Thanks for the PR, I have a suggestion.

abompard commented on noggin/templates/_agreements_form.html:
```diff
@@ -1,6 +1,21 @@
+      {# Determine whether this agreement is the FPCA #}
+      {% set is_fpca = agreement.slug == "fpca"
+                        or agreement.name|lower == "fpca"
+                        or agreement.name|lower.startswith("fedora project contributor agreement") %}
```
This doesn't work: 
`jinja2.exceptions.TemplateAssertionError: No filter named 'lower.startswith'.`

abompard reviewed (changes requested):
Thanks. I have one more comment and a suggestion: instead of hardcoding the name ("fpca" or "fedora project contributor agreement"), I think we should have a configuration value named, for example, `AGREEMENT_WARNINGS` that would be a dict where keys are agreement names and values are the warning message that should be displayed when the corresponding agreement is not signed.
This way we could make Noggin more independant from Fedora and it could be more easily reused by other projects.

Jkhall81 committed:
Add configurable agreement warnings and integrate FPCA warning into agreements form

Signed-off-by: Jkhall81 <jason.kei.hall@gmail.com>

abompard commented on noggin/defaults.py:
```diff
@@ -61,3 +61,11 @@
+# Agreement warnings shown when the corresponding agreement is not signed.
+# Keys must match the agreement.name from FreeIPA.
+AGREEMENT_WARNINGS = {
+    "Fedora Project Contributor Agreement": _(
```
For that to work you need to add `from flask_babel import _` at the top of the file.

abompard reviewed (changes requested):
Very nice, thanks!

For some ~stup~historic reason, the name of the FPCA agreement in the devel environment is `FPCA`, not `Fedora Project Contributor Agreement`. Could you add the same config block (with the `FPCA` key) to the `devel/ansible/roles/noggin/templates/noggin.cfg` config file as well please?

Also, please add a changelog fragment as described [in this doc](https://noggin-aaa.readthedocs.io/en/latest/contributing.html#release-notes).

Almost there! :-)

abompard commented on noggin/controller/registration.py:
```diff
@@ -355,6 +355,7 @@ def sign_agreements(ipa):
         user=user,
         agreementslist=agreements,
         all_signed=all(agreement.name in user.agreements for agreement in agreements),
+        agreement_warnings=current_app.config.get('AGREEMENT_WARNINGS', {}),
```
Since you've added the key to `defaults.py`, you know it's there, you can use:
```suggestion
        agreement_warnings=current_app.config["AGREEMENT_WARNINGS"],
```

abompard commented on noggin/defaults.py:
```diff
@@ -61,3 +61,11 @@
+# Keys must match the agreement.name from FreeIPA.
+AGREEMENT_WARNINGS = {
+    "Fedora Project Contributor Agreement": _(
+        "Not signing the FPCA will prevent you from logging in to many Fedora services such as Pagure, Fedora Discussion, and other contributor platforms."
```
This is flagged as too long a line by `tox -e lint`.

abompard commented on noggin/templates/registration-agreements.html:
```diff
@@ -13,6 +12,7 @@
         <div class="card">
             <div class="card-body">
               <h4 class="card-title">{{_("Optional agreement signing")}}</h4>
+              
```
This line has empty spaces.

Jkhall81 committed:
Add FPCA agreement warning and update configs (fixes #1530)

Signed-off-by: Jkhall81 <jason.kei.hall@gmail.com>

abompard commented on devel/ansible/roles/noggin/templates/noggin.cfg:
```diff
@@ -67,3 +67,11 @@ MAIL_SUPPRESS_SEND = True
 # Spam checking
 BASSET_URL = None
 SPAMCHECK_TOKEN_EXPIRATION = 60  # in minutes
+
+# Agreement warnings shown when the corresponding agreement is not signed.
+# Keys must match the agreement name from FreeIPA.
+AGREEMENT_WARNINGS = {
+    "FPCA": "Not signing the FPCA will prevent you from logging in to many Fedora "
+            "services such as Pagure, Fedora Discussion, and other contributor "
+            "platforms."
+}
```
This file will need the `from flask_babel import _` as well.

abompard commented on noggin/defaults.py:
```diff
@@ -61,3 +62,13 @@
+# Agreement warnings shown when the corresponding agreement is not signed.
+# Keys must match the agreement.name from FreeIPA.
+AGREEMENT_WARNINGS = {
+    "FPCA": _(
```
The default agreement name is `Fedora Project Contributor Agreement`.

abompard commented on noggin/defaults.py:
```diff
@@ -1,5 +1,6 @@
 # This file contains the default configuration values
 import socket
+from flask_babel import _
```
Black suggests adding a space before this line (you can check with `tox -e format`).

abompard reviewed (changes requested):
Just a few fixes left.
One slightly bigger thing: the `_agreements_form.html` template is also used by the `noggin.controller.user.user_settings_agreements()` view. As a result we get an error because the `agreement_warnings` template variable is not defined in this view. I see 2 ways of solving this:
- we define the variable in the `user_settings_agreements()` view as well
- instead of defining a variable in the `sign_agreements()` view, we use the fact that the app config is available in the templates directly in the `config` variable, and change `agreement_warnings` to `config["AGREEMENT_WARNINGS"]` in `_agreements_form.html`.

I would prefer the second option because it's less duplicated code.

You can check that is all works by running the unit tests with `tox -e py312-unittest`

Jkhall81 committed:
Finalized FPCA warning config, updated defaults and devel settings, ensured compatibility with user settings agreements.

Signed-off-by: jkhall81 <jason.kei.hall@gmail.com>

Jkhall81 commented:
Thanks for the clarification on the devel environment agreement name — I had misread that part earlier.

I’ve updated the configuration accordingly. I also set up tox and poetry locally so I could run the full test suite, and everything now passes.

`tox -e format` and `tox -e lint` are clean as well.

Let me know if anything else needs adjustment!

abompard commented:
Hey, thanks for the changes! My last comment is still relevant, about changing `agreement_warnings` to `config["AGREEMENT_WARNINGS"]` in `_agreements_form.html`.

Also, I've made a small fix to the translations in one of the files you touched, please rebase.

Jkhall81 commented:
I rebased..  there was only a change in `registration-agreements.html`.

Also in `_agreements_form.html` I had already updated this line:

```jinja
{% set warning = config["AGREEMENT_WARNINGS"].get(agreement.name) %}
````

This is the one you are referring to correct?

abompard commented on news/1530.feature:
```diff
@@ -0,0 +1 @@
+Add warning to agreements page explaining that not signing the FPCA prevents logging in to most Fedora services
```
This file seems to have line endings in dos format (`\r\n`), please convert it to unix format (`\n`).

abompard commented on noggin/controller/registration.py:
```diff
@@ -355,6 +355,7 @@ def sign_agreements(ipa):
         user=user,
         agreementslist=agreements,
         all_signed=all(agreement.name in user.agreements for agreement in agreements),
+        agreement_warnings=current_app.config["AGREEMENT_WARNINGS"],
```
I don't think you need this anymore, since you're now using the `config` template variable.

abompard commented on noggin/templates/registration-agreements.html:
```diff
@@ -13,12 +12,9 @@
         <div class="card">
             <div class="card-body">
               <h4 class="card-title">{{_("Optional agreement signing")}}</h4>
-              <p>{{ngettext(
-                "Hello %(name)s. Please sign the agreement below if you agree with it.",
-                "Hello %(name)s. Please sign the agreements below if you agree with them.",
-                agreementslist|length,
-                name=user.name
-              )}}</p>
+
+              <p>{{_("Hello %(name)s. Please sign the agreement(s) below if you agree with them.", name=user.name)}}</p>
+
```
Hmm no, please leave the plural-aware translation string in place.

abompard reviewed (changes requested):
Still a few fixes, thanks!

Jkhall81 committed:
Normalized line endings, removed ununsed agreement_warning, and restored ngettext pluralization.

Signed-off-by: jkhall81 <jason.kei.hall@gmail.com>

abompard commented:
OK, looks good! Please rebase on the current develop branch, there's been an update to a couple translated strings that will conflict, but it should be easy to merge since you're only changing quotes and newlines.

Jkhall81 commented:
Rebase completed! I pulled in the latest develop changes, resolved the template string conflicts, and force-pushed the updated branch. Let me know if anything else needs adjusting.

abompard reviewed (approved):
Looks good, thanks a lot for your contribution!

Jkhall81 commented:
Very cool.
